libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2303,11 +2303,22 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
 		functionType = dynamic_cast<FunctionType const*>(expressionType);
 		funcCallAnno.kind = FunctionCallKind::FunctionCall;
 
+		{
+			FunctionDefinition const* expressionDeclaration = nullptr;
+
+			if (auto memberAccess = dynamic_cast<MemberAccess const*>(&_functionCall.expression()))
+				expressionDeclaration = dynamic_cast<FunctionDefinition const*>(memberAccess->annotation().referencedDeclaration);
+			else if (auto identifier = dynamic_cast<Identifier const*>(&_functionCall.expression()))
+				expressionDeclaration = dynamic_cast<FunctionDefinition const*>(identifier->annotation().referencedDeclaration);
+
+			if (expressionDeclaration)
+				_functionCall.expression().annotation().calledDirectly = true;
+		}
+
 		// Purity for function calls also depends upon the callee and its FunctionType
 		funcCallAnno.isPure =
 			argumentsArePure &&
 			*_functionCall.expression().annotation().isPure &&
-			functionType &&
```

Comment:
If you remove this, please change the cast above to a `const&`-cast.

Reply:
I removed this because a few lines down no check is done anymore before accessing `functionType` so we might as well not have it here, too (or have it everywhere).

We can't easily change this because `functionType` is initialized outside the switch scope and only assigned here, for a reference that isn't possible..

Reply:
ok!